feat(elixir): adapt to new identity design

polvorin opened the pull request:
* Adapts to new identity design
* Get rid of the separate sidecar and vault ffi
* Use rustler for rust>elixir bindings on Identity/Credentials (implementations/elixir/ockam/ockly/native/ockly)
* Remove identity service and verifier service from ockam (those were only used by the "sidecar")

@SanjoDeundiak 
I think it's ok to merge into your branch.   There are some couple of missing things that I can iterate after,  mainly:
* There is no history change comparison on elixir (need to write the binding for that),  right now the function just  happily says histories match
* Examples using secure channels on examples/elixir/* need to be adapted
* There are some error conditions that can be better handled on the lib
* The interface with aws vault is there,  but I haven't tested it

[This](https://github.com/build-trust/ockam/blob/34a6f1ee2a960f213b6b53e38753f8eece02e892/implementations/elixir/ockam/ockly/native/ockly/src/lib.rs) is what replace the sidecar ,  some light :eyes: welcomed,  there is likely some generalization to be done there but we can iterate as well.   The interface is limited what was needed on elixir' side (so for example,  it only attest to securechannel keys,  the secret type it's not given as argument, etc).

SanjoDeundiak committed:
feat(rust): use `v2` identity in `ockam` crate

SanjoDeundiak committed:
feat(rust): use `v2` identity in `ockam_abac` crate

SanjoDeundiak committed:
feat(rust): use `v2` identity in `ockam_command` crate

SanjoDeundiak committed:
feat(rust): use `v2` identity in `ockam_api` crate

SanjoDeundiak committed:
feat(rust): extend `v2` identity api

SanjoDeundiak committed:
feat(rust): improve verify credential api

SanjoDeundiak committed:
feat(rust): extend `Identifier` api

SanjoDeundiak committed:
fix(rust): fix `Identity`-related builders

SanjoDeundiak committed:
feat(rust): improve `Identifier` serialization

SanjoDeundiak committed:
feat(rust): remove const-time comparison for `Identifier`

SanjoDeundiak committed:
chore(rust): fix clippy warnings

SanjoDeundiak committed:
feat(rust): replace `controller.id` with temporary value

SanjoDeundiak committed:
test(rust): replace fixture of enrollment ticket

SanjoDeundiak committed:
fix(rust): fix `get_started` compilation

SanjoDeundiak committed:
fix(rust): fix `ockam_app` compilation

SanjoDeundiak committed:
chore(rust): fix `get_started` examples

SanjoDeundiak committed:
feat(rust): add `import_private_identity` method

SanjoDeundiak committed:
feat(rust): check if public key matched on private identity import

SanjoDeundiak committed:
feat(rust): update ffi according to the new `Vault` design

SanjoDeundiak committed:
feat(rust): update `ockam_command` to new `Vault` structure

SanjoDeundiak committed:
feat(rust): update `ockam_api` to new `Vault` structure

SanjoDeundiak committed:
feat(rust): update `ockam` to new `Vault` structure

SanjoDeundiak committed:
feat(rust): disable old identity implementation to avoid updating it with new `Vault`

SanjoDeundiak committed:
test(rust): delete old identity implementation tests

SanjoDeundiak committed:
feat(rust): update `AWS` vault

SanjoDeundiak committed:
fix(rust): change p256 public key length to 65

SanjoDeundiak committed:
fix(rust): fix `import_private_identity` fn

SanjoDeundiak committed:
test(rust): replace all-zero test identity with a random

SanjoDeundiak committed:
test(rust): fix compilation for aws `ockam_command` test

SanjoDeundiak committed:
feat(rust): replace old identity implementation with the new one

SanjoDeundiak committed:
fix(rust): fix `no_std` build

SanjoDeundiak committed:
feat(rust): remove deprecated `PROJECT_ID` from credentials check

SanjoDeundiak committed:
chore(rust): decrease level of logging for some messages

SanjoDeundiak committed:
feat(rust): rewrite `Vault`

SanjoDeundiak committed:
feat(rust): update `ockam_identity` to new `Vault` structure

SanjoDeundiak committed:
feat(rust): save `Identity` to the storage during import

SanjoDeundiak committed:
test(rust): update kafka integration test

SanjoDeundiak committed:
test(rust): update bats test to support new `Identity`

SanjoDeundiak committed:
fix(rust): fix `expected_identifier` check during credentials verification

SanjoDeundiak committed:
feat(rust): add `encoding_utils` to `ockam_identity`

SanjoDeundiak committed:
feat(rust): improve display implementation for identity-related models

SanjoDeundiak committed:
feat(rust): add `create` to software vaults

SanjoDeundiak committed:
feat(rust): change `import_private_identity` to accept binary instead of hex

SanjoDeundiak committed:
feat(rust): implement and use lazy purpose key creation

SanjoDeundiak committed:
feat(rust): add sync interface for `SoftwareVerifyingVault`

SanjoDeundiak committed:
feat(rust): improve `ockam/src/node.rs`

SanjoDeundiak committed:
chore(rust): organize vault imports

SanjoDeundiak committed:
test(rust): fix secure_channel/vault integration test

SanjoDeundiak committed:
feat(rust): improve aes length checks

SanjoDeundiak committed:
feat(rust): remove `PurposeKeys` from `SecureChannels`

SanjoDeundiak committed:
feat(rust): improve `Identity` verification

SanjoDeundiak committed:
test(rust): improve `Identity` verification tests

SanjoDeundiak committed:
chore(rust): clean up identities `mod.rs`

SanjoDeundiak committed:
feat(rust): delete the old key during rotation

SanjoDeundiak committed:
feat(rust): add persistent `Identity` rotation

SanjoDeundiak committed:
feat(rust): improve `PurposeKey` and `Credential` verification

SanjoDeundiak committed:
feat(rust): add `TRUST_CONTEXT_ID_UTF8` for type-safety

SanjoDeundiak committed:
test(rust): fix `Identity` comparison test

SanjoDeundiak committed:
chore(rust): remove `extern crate core;` from `ockam_command`

SanjoDeundiak committed:
feat(rust): add key type check on identity creation with an existing key

polvorin committed:
feat(elixir): adapt to new identity design

Adapts to new identity desing, get rid of the separate sidecar and vault ffi.
Use rustler for rust>elixir bindings on Identity/Credentials.

SanjoDeundiak committed:
fix(rust): rebase fixes

SanjoDeundiak committed:
chore(rust): update import formats in `ockam_identity`

SanjoDeundiak committed:
feat(rust): add `IdentityBuilder`

SanjoDeundiak committed:
feat(rust): add `export_identity` and `get_identity` to `Identities`

SanjoDeundiak committed:
test(rust): fix aws identity test

SanjoDeundiak committed:
feat(rust): split `PurposeKeys` into creation and verification

SanjoDeundiak committed:
feat(rust): split `Credentials` into creation and verification

SanjoDeundiak committed:
feat(rust): move models conversions

SanjoDeundiak committed:
feat(rust): add `PurposeKeyBuilder`

SanjoDeundiak committed:
feat(rust): improve `IdentityBuilder` and `PurposeKeyBuilder`

SanjoDeundiak committed:
feat(rust): better key type checks

SanjoDeundiak committed:
test(rust): improve identity and purposekey tests

SanjoDeundiak committed:
chore(rust): no_std fix

SanjoDeundiak committed:
chore(rust): add a `FIXME` for `TrustContext`

SanjoDeundiak committed:
feat(rust): split identity models into multiple files

SanjoDeundiak committed:
feat(rust): split signing_vault into identity_vault and credential_vault

SanjoDeundiak committed:
feat(rust): move `Vault` structure to `ockam_identity`

SanjoDeundiak committed:
feat(rust): optimize credentials on-the-wire format

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/nodes/service/node_identities.rs:
```diff
@@ -27,47 +24,11 @@ impl NodeIdentities {
         self.identities.vault()
     }
 
-    pub(super) fn identities_repository(&self) -> Arc<dyn IdentitiesRepository> {
```
I'm not sure removing it is useful and related to that PR, but if these functions just become unused, should be OK

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+    .map_err(|_| Error::BadArg)?;
+
+    let exported = identity.export().map_err(|_| Error::BadArg)?;
+    let id = identity.identifier().to_string();
```
Let's keep it in binary as 20 bytes

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+
+    let exported = identity.export().map_err(|_| Error::BadArg)?;
+    let id = identity.identifier().to_string();
+    let mut binary = NewBinary::new(env, id.len());
```
It's surprising how inconvenient it is. `&[u8]` -> `NewBinary` could have been a one simple call instead of 3

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+#[rustler::nif]
+fn attest_purpose_key<'a>(
+    env: Env<'a>,
+    identifier: String,
```
Let's use binaries where we can

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+            .vault()
+            .secure_channel_vault
+            .import_static_secret(
+                Secret::new(public_key.to_vec()),
```
That doesn't look correct. I will add a possibility to attest a PurposeKey having only a public key available.

polvorin replied:
oh,  good catch, the *name* is wrong!.    The argument is not the `public_key`,  it's actually the `secret` .     It is hacky as we import the secret just to be able to attest to the corresponding public key.  

Originally I added what you said (a way to attest to public key directly) and so the public key was passed,  but removed that to not introduce changes on that part of the ockam_identity crate that was giving me conflicts.   But forgot to rename the argument according.     Will fix it (and yes not requiring the secret will simplify this part, will be a good addition)

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+    block_future(async move {
+        let identity = identities_ref
+            .identities_creation()
+            .import(None, &identity)
```
Are the cases where we should check the Identifier match here or on the calling side?

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+}
+
+#[rustler::nif]
+fn verify_purpose_key_attestation(
```
Let's rename it since it only works with Secure Channel key. Or let's add support for Credentials keys

polvorin replied:
right,   this and the above one are only for secure channels keys, will rename

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+fn issue_credential<'a>(
+    env: Env<'a>,
+    issuer_identity: Binary,
+    subject_identifier: String,
```
Do we import subject's Identity before making that call?

polvorin replied:
ah,  i see what you mean.   We don't enforce that,  it happens to be that way because this is used for tests so the identity is generated by us.    But will fix this so it's evident what's going on

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+            .import(None, &issuer_identity)
+            .await
+            .map_err(|_| atoms::identity_import_error())?;
+        let mut attr_builder = AttributesBuilder::with_schema(SchemaId(0));
```
Schema should be customizable I guess

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+#[rustler::nif]
+fn issue_credential<'a>(
+    env: Env<'a>,
+    issuer_identity: Binary,
```
Issuer should have valid Credential PurposeKey to perform that, or the default one will be created under the hood

polvorin replied:
yes,  I assumed a default one would be created (and a new default after a restart),  but it's ok this is for testing only right now  (we don't have an authority implementation on elixir,   so only need to verify credentials but not really generate them other than for testing/examples)

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+    let identities_ref = identities_ref()?;
+    let subject_identifier = Identifier::from_str(&subject_identifier)
+        .map_err(|_| Error::Term(Box::new(atoms::invalid_identifier())))?;
+    let credential = block_future(async move {
```
That's also `CredentialAndPurposeKey` FYI

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+#[rustler::nif]
+fn verify_credential(
+    expected_subject: String,
+    authorities: Vec<Binary>,
```
I wonder if it's easier to first import Identities, and then pass only Identifiers. Not only for this call, for others as well

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+            let builder = ockam_identity::Identities::builder().with_vault(Vault::new(
+                Arc::new(vault),
+                Vault::create_secure_channel_vault(),
+                Vault::create_credential_vault(),
```
This one should be AWS as well

polvorin replied:
thanks!, will fix

SanjoDeundiak commented on implementations/elixir/ockam/ockly/native/ockly/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+}
+
+#[rustler::nif]
+fn import_signing_secret(secret: Binary) -> NifResult<String> {
```
Is it required?

SanjoDeundiak reviewed (changes requested):
I didn't review Elixir code. Also, I haven't had experience with rustler, so may be missing some details on correct use of it. I reviewed rust code, looks like a good start. We can merge it and iterate from there

SanjoDeundiak commented:
Could you please look at why CI fails? I think only fails that are already present in my PR are expected here, others can be fixed.

polvorin committed:
feat(elixir): identifiers as raw binaries

Some API expect identities as raw binaries (others expect
them as hex strings). Provide some workaround around that.

polvorin commented:
Closing in favor of https://github.com/build-trust/ockam/pull/5976.   Created issues for the pending TODOs
